# Оглавление
1. [Сингулярное разложение матриц](#svd)
    * [Задание 1](#svd_1)
    * [Задание 2](#svd_2)

In [1]:
import numpy as np
np.set_printoptions(precision=2, suppress=True)

# Сингулярное разложение матриц <a class='anchor' id='svd'>

### Задание 1 <a class='anchor' id='svd_1'>
    
Найдите посредством NumPy SVD-разложение для матрицы:

$$
\begin{pmatrix}
1 & 2 & 0 \\ 
0 & 0 & 5 \\ 
3 & -4 & 2 \\ 
1 & 6 & 5 \\ 
0 & 1 & 0
\end{pmatrix}.
$$
    
**Решение**

In [2]:
# Исходная матрица
A = np.array([[1, 2, 0], [0, 0, 5], [3, -4, 2], [1, 6, 5], [0, 1, 0]])

# SVD-разложение
U, s, W = np.linalg.svd(A)

# Транспонируем матрицу W
V = W.T

# D - матрица, на диоганали которой расположены значения из s - список сингулярных чисел
D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s

print(f'A =\n{A}\n')
print(f'U =\n{U}\n')
print(f'D =\n{D}\n')
print(f'V =\n{V}\n')
print(f'Сингулярные числа: {s}')

A =
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]

U =
[[ 0.17  0.16 -0.53 -0.8  -0.16]
 [ 0.39 -0.53  0.61 -0.43  0.03]
 [-0.14 -0.82 -0.52  0.14  0.07]
 [ 0.89  0.06 -0.25  0.38 -0.06]
 [ 0.08  0.11 -0.08 -0.11  0.98]]

D =
[[8.82 0.   0.  ]
 [0.   6.14 0.  ]
 [0.   0.   2.53]
 [0.   0.   0.  ]
 [0.   0.   0.  ]]

V =
[[ 0.07 -0.37 -0.93]
 [ 0.72  0.67 -0.21]
 [ 0.69 -0.65  0.31]]

Сингулярные числа: [8.82 6.14 2.53]


In [3]:
# Проверим, что умножение полученных матриц дает исходную матрицу A
U @ D @ V.T

array([[ 1.,  2.,  0.],
       [ 0., -0.,  5.],
       [ 3., -4.,  2.],
       [ 1.,  6.,  5.],
       [-0.,  1., -0.]])

### Задание 2 <a class='anchor' id='svd_2'>
    
Для матрицы из предыдущего задания найдите:

а) евклидову норму;
    
б) норму Фробениуса.
    
**Решение**
    
Евклидова норма матрицы равна максимальному сингулярному числу:
    
$$\left \| A \right \|_E = \mu_1 = 8.82.$$
    
Норма Форбениуса с использованием сингулярных чисел вычисляется следующим образом:
    
$$\left \| A \right \|_F = \sqrt{\sum_{k = 1}^r \mu_k^{2}} = 11.045.$$

In [4]:
# Вычисление нормы Фробениуса
norm_F = np.linalg.norm(s)
print(f'Норма Фробениуса = {norm_F}\n')

Норма Фробениуса = 11.045361017187263



In [5]:
# Проверка расчета нормы
norm_E_np = np.linalg.norm(A, ord=2)
norm_F_np = np.linalg.norm(A, ord='fro')

print('Проверка через numpy')
print(f'Евклидова норма = {norm_E_np}')
print(f'Норма Фробениуса = {norm_F_np}')

Проверка через numpy
Евклидова норма = 8.824868854820444
Норма Фробениуса = 11.045361017187261


**Ответ (задание 2):** 

$\left \| A \right \|_E = 8.82$

$\left \| A \right \|_F = 11.045$